Source/Meadow.Core/Cloud/MeadowCloudUpdateService.cs

```diff
@@ -230,31 +231,56 @@ private async Task DownloadProc(UpdateMessage message)
                     {
-                        httpClient.DefaultRequestHeaders.Authorization = _connectionService.CreateAuthenticationHeaderValue();
-                    }
-
-                    // Configure the HTTP range header to indicate resumption of partial download, starting from 
-                    // the 'totalBytesDownloaded' byte position and extending to the end of the content.
-                    httpClient.DefaultRequestHeaders.Range = new System.Net.Http.Headers.RangeHeaderValue(totalBytesDownloaded, null);
+                        if (_connectionService.Settings.UseAuthentication)
+                        {
+                            request.Headers.Authorization = _connectionService.CreateAuthenticationHeaderValue();
+                        }
 
-                    using (var stream = await httpClient.GetStreamAsync(destination))
-                    {
-                        using (var fileStream = Store.GetUpdateFileStream(message.ID))
+                        // Get the actual update file size, which might be larger than the expected
+                        // if this OTA requires an OS update
+                        var response = await httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead);
+                        var contentLength = response.Content.Headers.ContentLength;
+                        if (contentLength.HasValue)
+                        {
+                            message.FileSize = contentLength.Value;
+                            Resolver.Log.Trace($"Starting download... File size: {message.FileSize.ToString("N0")} bytes.");
+                        }
+                        else
                         {
-                            byte[] buffer = new byte[1024 * 64]; // TODO: make this configurable/platform dependent
-                            int bytesRead;
+                            Resolver.Log.Warn($"Starting download... File size is unknown. Assuming it is {message.FileSize.ToString("N0")} bytes.");
+                        }
 
-                            while ((bytesRead = await stream.ReadAsync(buffer, 0, buffer.Length)) > 0)
+                        // Configure the HTTP range header to indicate resumption of partial download, starting from 
+                        // the 'totalBytesDownloaded' byte position and extending to the end of the content.
+                        if (totalBytesDownloaded <= message.FileSize)
+                        {
+                            request.Headers.Range = new System.Net.Http.Headers.RangeHeaderValue(totalBytesDownloaded, null);
+                        }
+                        else
+                        {
+                            // Ignore the total bytes downloaded previously and start a new download again
+                            Resolver.Log.Warn($"Total bytes downloaded ({totalBytesDownloaded.ToString("N0")}) exceed the file size ({message.FileSize.ToString("N0")}).");
+                        }
```

Comment:
An issue I see with this is that the HTTP request has already been sent on line 243. I believe that the Range header won't have any effect here.

Reply:
Yeah, thanks! I need to adjust it given that now we are making the request on line 243 indeed.